expression/builtin_regexp_util.go

```diff
@@ -0,0 +1,213 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expression
+
+import (
+	"regexp"
+	"sync"
+
+	"github.com/pingcap/tidb/util/chunk"
+)
+
+// Parameters may be const or ignored by the user, so different situations should be considered
+// We can handle parameters more easily with this struct.
+//
+// notProvided field shows if user ignores this param in sql
+// for example:
+//
+//	select regexp_like("123", "123", "m"), here notProvided field is false for the third parameter
+//	select regexp_like("123", "123"), here notProvided field is true for the third parameter
+type regexpParam struct {
+	constStrVal   string
+	constIntVal   int64
+	defaultIntVal int64 // default value when notProvided is true
+	isConst       bool
+	notProvided   bool
+	col           *chunk.Column
+}
+
+func (re *regexpParam) getCol() *chunk.Column {
+	return re.col
+}
+
+func (re *regexpParam) getStringVal(id int) string {
+	if re.notProvided {
+		return ""
+	}
+
+	if re.isConst {
+		return re.constStrVal
+	}
+
+	return re.getCol().GetString(id)
+}
+
+func (re *regexpParam) getIntVal(id int) int64 {
+	if re.notProvided {
+		return re.defaultIntVal
+	}
+
+	if re.isConst {
+		return re.constIntVal
+	}
+
+	return re.getCol().GetInt64(id)
+}
+
+// bool return value: return true when we get a const null parameter
+func buildStringParam(bf *baseBuiltinFunc, id int, input *chunk.Chunk, notProvided bool) (*regexpParam, bool, error) {
+	var pa regexpParam
+	var err error
+
+	pa.notProvided = notProvided
+	if pa.notProvided {
+		return &pa, false, nil
+	}
+
+	pa.col, err = bf.bufAllocator.get()
+	if err != nil {
+		return nil, false, err
+	}
+
+	// Get values from input
+	if err := bf.args[id].VecEvalString(bf.ctx, input, pa.getCol()); err != nil {
+		return nil, false, err
+	}
+
+	// Check if this is a const value
+	pa.isConst = bf.args[id].ConstItem(bf.ctx.GetSessionVars().StmtCtx)
+	if pa.isConst {
+		// Initialize the const
+		var isConstNull bool
+		pa.constStrVal, isConstNull = getColumnConstValString(pa.getCol(), input.NumRows())
+		if isConstNull {
+			return nil, isConstNull, nil
+		}
+	}
+
+	return &pa, false, nil
+}
+
+// bool return value: return true when we get a const null parameter
+func buildIntParam(bf *baseBuiltinFunc, id int, input *chunk.Chunk, notProvided bool, defaultIntVal int64) (*regexpParam, bool, error) {
+	var pa regexpParam
+	var err error
+
+	pa.notProvided = notProvided
+	if pa.notProvided {
+		pa.defaultIntVal = defaultIntVal
+		return &pa, false, nil
+	}
+
+	pa.col, err = bf.bufAllocator.get()
+	if err != nil {
+		return nil, false, err
+	}
+
+	// Get values from input
+	if err := bf.args[id].VecEvalInt(bf.ctx, input, pa.getCol()); err != nil {
+		return nil, false, err
+	}
+
+	// Check if this is a const value
+	pa.isConst = bf.args[id].ConstItem(bf.ctx.GetSessionVars().StmtCtx)
```

Comment:
Should we check `pa.isConst` before `init pa.col`?

Reply:
> Should we check `pa.isConst` before `init pa.col`?

yes, I will change it

Reply:
> Should we check `pa.isConst` before `init pa.col`?

nope, we need pa.col when initializing the const

Reply:
> Should we check `pa.isConst` before `init pa.col`?

done